README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# <img src="https://i.imgur.com/m8FNhQR.png" align="right" height=88 /> Analize soils and tree-habitat associations

[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![Travis build
status](https://travis-ci.org/forestgeo/fgeo.habitat.svg?branch=master)](https://travis-ci.org/forestgeo/fgeo.habitat)
[![Coverage
status](https://coveralls.io/repos/github/forestgeo/fgeo.habitat/badge.svg)](https://coveralls.io/r/forestgeo/fgeo.habitat?branch=master)
[![CRAN
status](http://www.r-pkg.org/badges/version/fgeo.habitat)](https://cran.r-project.org/package=fgeo.habitat)

## Installation

``` r
# install.packages("remotes")
remotes::install_github("forestgeo/fgeo.habitat")
```

For details on how to install packages from GitHub, see [this
article](https://goo.gl/dQKEeg).

## Example

``` r
library(fgeo.habitat)
# For easier data wranging
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
```

### Habitat-species associations.

``` r
# Example data
habitat <- luquillo_habitat
census <- luquillo_top3_sp

# Pick alive trees, of 10 mm or more
pick <- filter(census, status == "A", dbh >= 10)
# Pick sufficiently abundant trees
pick <- add_count(pick, sp)
pick <- filter(pick, n > 50)

species <- unique(pick$sp)

tt <- tt_test(census, species, habitat)
tt
#> [[1]]
#>        N.Hab.1 Gr.Hab.1 Ls.Hab.1 Eq.Hab.1 Rep.Agg.Neut.1 Obs.Quantile.1
#> CASARB      25     1489      109        2              0       0.930625
#>        N.Hab.2 Gr.Hab.2 Ls.Hab.2 Eq.Hab.2 Rep.Agg.Neut.2 Obs.Quantile.2
#> CASARB      12      168     1431        1              0          0.105
#>        N.Hab.3 Gr.Hab.3 Ls.Hab.3 Eq.Hab.3 Rep.Agg.Neut.3 Obs.Quantile.3
#> CASARB      14      567     1029        4              0       0.354375
#>        N.Hab.4 Gr.Hab.4 Ls.Hab.4 Eq.Hab.4 Rep.Agg.Neut.4 Obs.Quantile.4
#> CASARB      15      934      661        5              0        0.58375
#> 
#> [[2]]
#>        N.Hab.1 Gr.Hab.1 Ls.Hab.1 Eq.Hab.1 Rep.Agg.Neut.1 Obs.Quantile.1
#> PREMON      59      389     1208        3              0       0.243125
#>        N.Hab.2 Gr.Hab.2 Ls.Hab.2 Eq.Hab.2 Rep.Agg.Neut.2 Obs.Quantile.2
#> PREMON      75     1562       37        1              1        0.97625
#>        N.Hab.3 Gr.Hab.3 Ls.Hab.3 Eq.Hab.3 Rep.Agg.Neut.3 Obs.Quantile.3
#> PREMON      56      632      963        5              0          0.395
#>        N.Hab.4 Gr.Hab.4 Ls.Hab.4 Eq.Hab.4 Rep.Agg.Neut.4 Obs.Quantile.4
#> PREMON      44      222     1375        3              0        0.13875
#> 
#> [[3]]
#>        N.Hab.1 Gr.Hab.1 Ls.Hab.1 Eq.Hab.1 Rep.Agg.Neut.1 Obs.Quantile.1
#> SLOBER      14      492     1092       16              0         0.3075
#>        N.Hab.2 Gr.Hab.2 Ls.Hab.2 Eq.Hab.2 Rep.Agg.Neut.2 Obs.Quantile.2
#> SLOBER      16      473     1125        2              0       0.295625
#>        N.Hab.3 Gr.Hab.3 Ls.Hab.3 Eq.Hab.3 Rep.Agg.Neut.3 Obs.Quantile.3
#> SLOBER      19     1181      415        4              0       0.738125
#>        N.Hab.4 Gr.Hab.4 Ls.Hab.4 Eq.Hab.4 Rep.Agg.Neut.4 Obs.Quantile.4
#> SLOBER      17     1151      440        9              0       0.719375
#> 
#> attr(,"class")
#> [1] "tt_lst" "list"
```

``` r
tt_df <- to_df(tt)
# Try also: View(tt_df)
head(tt_df)
#>           metric     sp       value
#> 1        N.Hab.1 CASARB   25.000000
#> 2       Gr.Hab.1 CASARB 1489.000000
#> 3       Ls.Hab.1 CASARB  109.000000
#> 4       Eq.Hab.1 CASARB    2.000000
#> 5 Rep.Agg.Neut.1 CASARB    0.000000
#> 6 Obs.Quantile.1 CASARB    0.930625

tail(tt_df)
#>            metric     sp       value
#> 67        N.Hab.4 SLOBER   17.000000
#> 68       Gr.Hab.4 SLOBER 1151.000000
#> 69       Ls.Hab.4 SLOBER  440.000000
#> 70       Eq.Hab.4 SLOBER    9.000000
#> 71 Rep.Agg.Neut.4 SLOBER    0.000000
#> 72 Obs.Quantile.4 SLOBER    0.719375
```

### Krige soil data.

``` r
result <- krig(soil_fake, var = c("c", "p"), quiet = TRUE)
head(to_df(result))
#>   var   x  y        z
#> 1   c  10 10 2.134696
#> 2   c  30 10 2.119651
#> 3   c  50 10 2.104591
#> 4   c  70 10 2.089517
#> 5   c  90 10 2.074427
#> 6   c 110 10 2.059322

tail(to_df(result))
#>      var   x   y        z
#> 2295   p 890 450 5.835048
#> 2296   p 910 450 5.826698
#> 2297   p 930 450 5.819219
#> 2298   p 950 450 5.812612
#> 2299   p 970 450 5.806876
#> 2300   p 990 450 5.802012
```

``` r
summary(result)
#> var: c 
#> df
#> 'data.frame':    1150 obs. of  3 variables:
#>  $ x: num  10 30 50 70 90 110 130 150 170 190 ...
#>  $ y: num  10 10 10 10 10 10 10 10 10 10 ...
#>  $ z: num  2.13 2.12 2.1 2.09 2.07 ...
#> 
#> df.poly
#> 'data.frame':    1150 obs. of  3 variables:
#>  $ gx: num  10 30 50 70 90 110 130 150 170 190 ...
#>  $ gy: num  10 10 10 10 10 10 10 10 10 10 ...
#>  $ z : num  2.13 2.12 2.1 2.09 2.07 ...
#> 
#> lambda
#> 'numeric'
#>  num 1
#> 
#> vg
#> 'variogram'
#> List of 20
#>  $ u               : num [1:9] 60.9 86.5 103 122.7 146.1 ...
#>  $ v               : num [1:9] 0.284 0.422 0.882 0.543 0.211 ...
#>  $ n               : num [1:9] 7 9 10 10 18 19 36 34 38
#>  $ sd              : num [1:9] 0.414 0.48 0.633 0.501 0.405 ...
#>  $ bins.lim        : num [1:31] 1.00e-12 2.00 2.38 2.84 3.38 ...
#>  $ ind.bin         : logi [1:30] FALSE FALSE FALSE FALSE FALSE FALSE ...
#>  $ var.mark        : num 0.317
#>  $ beta.ols        : num 1.36e-09
#>  $ output.type     : chr "bin"
#>  $ max.dist        : num 320
#>  $ estimator.type  : chr "classical"
#>  $ n.data          : int 30
#>  $ lambda          : num 1
#>  $ trend           : chr "cte"
#>  $ pairs.min       : num 5
#>  $ nugget.tolerance: num 1e-12
#>  $ direction       : chr "omnidirectional"
#>  $ tolerance       : chr "none"
#>  $ uvec            : num [1:30] 1 2.19 2.61 3.11 3.7 ...
#>  $ call            : language variog(geodata = geodata, breaks = breaks, trend = trend, pairs.min = 5)
#> 
#> vm
#> 'variomodel', variofit'
#> List of 17
#>  $ nugget               : num 0.352
#>  $ cov.pars             : num [1:2] 0 160
#>  $ cov.model            : chr "exponential"
#>  $ kappa                : num 0.5
#>  $ value                : num 4.64
#>  $ trend                : chr "cte"
#>  $ beta.ols             : num 1.36e-09
#>  $ practicalRange       : num 480
#>  $ max.dist             : num 320
#>  $ minimisation.function: chr "optim"
#>  $ weights              : chr "npairs"
#>  $ method               : chr "WLS"
#>  $ fix.nugget           : logi FALSE
#>  $ fix.kappa            : logi TRUE
#>  $ lambda               : num 1
#>  $ message              : chr "optim convergence code: 0"
#>  $ call                 : language variofit(vario = vg, ini.cov.pars = c(initialVal, startRange), cov.model = varModels[i],      nugget = initialVal)
#> 
#> var: p 
#> df
#> 'data.frame':    1150 obs. of  3 variables:
#>  $ x: num  10 30 50 70 90 110 130 150 170 190 ...
#>  $ y: num  10 10 10 10 10 10 10 10 10 10 ...
#>  $ z: num  6.37 6.35 6.33 6.31 6.29 ...
#> 
#> df.poly
#> 'data.frame':    1150 obs. of  3 variables:
#>  $ gx: num  10 30 50 70 90 110 130 150 170 190 ...
#>  $ gy: num  10 10 10 10 10 10 10 10 10 10 ...
#>  $ z : num  6.37 6.35 6.33 6.31 6.29 ...
#> 
#> lambda
#> 'numeric'
#>  num 1
#> 
#> vg
#> 'variogram'
#> List of 20
#>  $ u               : num [1:9] 60.9 86.5 103 122.7 146.1 ...
#>  $ v               : num [1:9] 0.396 0.4 0.11 0.402 0.385 ...
#>  $ n               : num [1:9] 7 9 10 10 18 19 36 34 38
#>  $ sd              : num [1:9] 0.592 0.414 0.133 0.409 0.488 ...
#>  $ bins.lim        : num [1:31] 1.00e-12 2.00 2.38 2.84 3.38 ...
#>  $ ind.bin         : logi [1:30] FALSE FALSE FALSE FALSE FALSE FALSE ...
#>  $ var.mark        : num 0.267
#>  $ beta.ols        : num -3.14e-09
#>  $ output.type     : chr "bin"
#>  $ max.dist        : num 320
#>  $ estimator.type  : chr "classical"
#>  $ n.data          : int 30
#>  $ lambda          : num 1
#>  $ trend           : chr "cte"
#>  $ pairs.min       : num 5
#>  $ nugget.tolerance: num 1e-12
#>  $ direction       : chr "omnidirectional"
#>  $ tolerance       : chr "none"
#>  $ uvec            : num [1:30] 1 2.19 2.61 3.11 3.7 ...
#>  $ call            : language variog(geodata = geodata, breaks = breaks, trend = trend, pairs.min = 5)
#> 
#> vm
#> 'variomodel', variofit'
#> List of 17
#>  $ nugget               : num 0.305
#>  $ cov.pars             : num [1:2] 0 160
#>  $ cov.model            : chr "exponential"
#>  $ kappa                : num 0.5
#>  $ value                : num 0.818
#>  $ trend                : chr "cte"
#>  $ beta.ols             : num -3.14e-09
#>  $ practicalRange       : num 479
#>  $ max.dist             : num 320
#>  $ minimisation.function: chr "optim"
#>  $ weights              : chr "npairs"
#>  $ method               : chr "WLS"
#>  $ fix.nugget           : logi FALSE
#>  $ fix.kappa            : logi TRUE
#>  $ lambda               : num 1
#>  $ message              : chr "optim convergence code: 0"
#>  $ call                 : language variofit(vario = vg, ini.cov.pars = c(initialVal, startRange), cov.model = varModels[i],      nugget = initialVal)
```

## Information

  - [Getting help](SUPPORT.md).
  - [Contributing](CONTRIBUTING.md).
  - [Contributor Code of Conduct](CODE_OF_CONDUCT.md).

## Acknowledgements

Thanks to all partners of ForestGEO who shared their ideas and code.
Functions’ authors include Graham Zemunik, Sabrina Russo, Daniel Zuleta,
Matteo Detto, Kyle Harms, Gabriel Arellano.


NEWS.md
# fgeo.habitat 0.0.0.9006



# fgeo.habitat 0.0.0.9005

This release removes non essential components to reduce complexity.

* Remove class tt and krig.
* Remove krig_lst() and tt_test_lst().
* Remove as_*().

# fgeo.habitat 0.0.0.9004

* Add a `summary()` method for class "krig".
* Add generic `to_df()` for classes "tt", "tt_lst" and "krig_lst".
* New generic to coerce results of `GetKrigedSoil()` to "krig" and "krig_lst".
* New generic to coerce results of `torusonesp.all()` to "tt" and "tt_lst".
* Split vignette fgeo.habitat.Rmd into tt_test.Rmd and krig.Rmd.

# fgeo.habitat 0.0.0.9003

* Improve performance of tt_test() and abund_index().

# fgeo.habitat 0.0.0.9002

* New function `tt_test()` and friends, by Sabrina Russo et al.
* Rename to __fgeo.habitat__

# soilkrig 0.0.0.9000

* Update website for consistency with other packages of __fgeo__
* Added a `NEWS.md` file to track changes to the package.


R/krig_lst.R
#' Krige soil data following the methodology of the John et al. (2007).
#'
#' Functions to krige soil data:
#' * `GetKrigedSoil()` is the original function but it is softly deprecated. Its
#' result is the same as `krig()[[1]]` when run for one soil variable.
#' * `krig()` some additional convenient features:
#'     * An interface consistent with other functions of ForestGEO.
#'     * A a method for `summary()`(see examples).
#'     * Tries to guess `plotdim`.
#'     * Informs the guessed `plotdim` and the `gridsize` provided.
#'     * Aallows to  suppress messages.
#'     * Iterating over multiple soil variables.
#'     * Has a method for `to_df()` (see examples).
#' 
#' @inheritSection krig_auto_params Breaks default
#' 
#' @param soil,df.soil The data frame with the points, coords specified in the
#'   columns `gx`, `gy`.
#' @param var A character vector giving the name in the soil dataset of the
#'   variable(s) containing soil data to krige. `var` must be of length 1 for
#'   `GetKrigedSoil()` but can be of length greater than one for `krig()`.
#' @param gridsize,gridSize Points are kriged to the center points of a grid of
#'   this size.
#' @param params,krigeParams If you want to pass specified kriging parameters;
#'   see [krig_auto_params()] for each parameter.
#' @param plotdim,xSize,ySize Numeric vectors giving x and y dimensions of the
#'   plot: 
#'   * `plotdim`: Must be of length 2 with the format `c(x, y)`.
#'   * `xSize`, `ySize`: Each must be of length 1.
#' @param breaks Breaks/intervals used to calculate the semivariogram, which
#'   only happens if `krigeParams = NULL` (default).
#' @param use_ksline,useKsLine Use the [geoR::ksline()] function? Use `TRUE` to
#'   calculate a "best" semivariogram based on default parameters via
#'   `geoR::variogram()`]. Use `FALSE` to base calculation on parameters passed
#'   to `params`.
#' @param quiet Use `TRUE` to suppresses messages.
#' 
#' @return A list with the following items:
#'   * `df`: Data frame of kriged values (column z) at each grid point (x, y).
#'   * `df.poly`: Data frame of the polynomial surface fitted to the raw data.
#'   * `lambda`: The "lambda" value used in the Box-Cox transform of the raw
#'     data.
#'   * `vg`: A list giving the variogram parameters used for the kriging.
#'   * `vm`: Minimum loss value returned from [geoR::variofit()].
#' @seealso [geoR::variofit()], [geoR::variog()], [geoR::as.geodata()],
#'   [geoR::ksline()], [geoR::krige.conv()], [geoR::krige.control()].
#'
#' @author Graham Zemunik (grah.zem@@gmail.com).
#'
#' @export
#' @seealso [to_df.krig_lst()], [summary.krig_lst()].
#' 
#' @examples
#' # Original funciton
#' # Using automated parameters
#' auto <- GetKrigedSoil(soil_fake, var = "c")
#' summary(auto)
#' 
#' # Wrapper produces the same output and add convenient features
#' 
#' # Now using custom parameters (arbitrary but based on automated kriging params)
#' params <- list(
#'   model = "circular", range = 100, nugget = 1000, sill = 46000, kappa = 0.5
#' )
#' 
#' # Also using not one but multiple soil variables
#' vars <- c("c", "p")
#' custom <- krig(soil_fake, vars, params = params, quiet = TRUE)
#' summary(custom)
#' 
#' result <- to_df(custom, item = "df")
#' head(result)
#' 
#' tail(result)
krig <- function(soil,
                     var,
                     params = NULL,
                     gridsize = 20,
                     plotdim = guess_plotdim(soil),
                     breaks = krig_breaks(2, 320, 30),
                     use_ksline = TRUE,
                     quiet = FALSE) {
  force(var)
  out <- lapply(
    var, 
    krig_one, soil = soil, params = params, gridsize = gridsize,
    plotdim = plotdim, breaks = breaks, use_ksline = use_ksline, quiet = quiet
  )
  out <- stats::setNames(out, var)
  new_krig_lst(out)
}

#' @rdname krig
#' @export
GetKrigedSoil <- function(df.soil,
                          var,
                          gridSize = 20,
                          krigeParams = NULL,
                          xSize = 1000,
                          ySize = 500,
                          breaks = krig_breaks(2, 320, 30),
                          useKsLine = TRUE) {
  check_krig(
    df.soil = df.soil, var = var, gridSize = gridSize, xSize = xSize,
    ySize = ySize, breaks = breaks, krigeParams = krigeParams,
    useKsLine = useKsLine
  )

  df <- df.soil[, c("gx", "gy", var)]
  names(df)[3] <- "z"

  # This follows a similar methodology to the John 2007 paper:
  # 1. the data is transformed based on the optimal boxcox transform
  # 2. a polynomial regression curve is fitted - a second order
  #    polynomial is used
  # 3. the residuals from the polynomial fit are kriged: two steps, a)
  #    get a variogram, b) do the kriging
  # 4. the kriged values are added to the polynomial fit and they
  #    are backtranformed as required

  # The lambda for the box-cox transform is restricted to 0, 0.5 and 1.
  # Data with 0's in there are handled by the addition of a small constant
  # in the regression
  bc <- BoxCoxTransformSoil(df)
  df <- bc$df

  polyfit <- GetPolynomialFit(
    df,
    gridSize = gridSize, xSize = xSize, ySize = ySize
  )

  # Get the variogram parameters
  # If a polynomial fit was possible then the parameters come from
  # the residuals of the polynomial fit
  # Otherwise, the raw data is used (a failure to fit the polynomial
  # most likely indicates that no trend exists, so trend removal isn't
  # necessary)
  if (is.null(polyfit$mod)) {
    df.krig <- df
  } else {
    df.krig <- cbind(polyfit$df.orig[, c("gx", "gy")], z = resid(polyfit$mod))
  }

  geod <- as.geodata(df.krig)
  if (is.null(krigeParams)) {
    params <- krig_auto_params(geod, breaks = breaks)
  } else {
    params <- krigeParams
    if (!("kappa" %in% names(params))) {
      params$kappa <- 0
    }
  }

  # Do the kriging
  if (useKsLine) {
    krig <- krig_ksline(
      geod,
      locations = polyfit$df.interpolated[, c("gx", "gy")],
      cov.pars = c(params$sill, params$range),
      cov.model = params$model,
      nugget = params$nugget,
      kappa = params$kappa,
      lambda = 1
    )
  } else {
    krig <- krige.conv(
      geod,
      locations = polyfit$df.interpolated[, c("gx", "gy")],
      krige = krige.control(
        cov.pars = c(params$sill, params$range),
        cov.model = params$model,
        nugget = params$nugget,
        kappa = params$kappa,
        lambda = 1,
        aniso.pars = NULL
      )
    )
  }
  # Add the kriged results to the trend if necessary
  df.pred <- polyfit$df.interpolated
  if (is.null(polyfit$mod)) {
    df.pred$z <- krig$predict
  } else {
    df.pred$z <- df.pred$z + krig$predict
  }

  # Back transform (if required)
  df.pred <- InvBoxCoxTransformSoil(df.pred, bc$lambda, bc$delta)

  names(df.pred) <- c("x", "y", "z")

  # Return all useful data
  list(
    df = df.pred,
    df.poly = polyfit$df.interpolated,
    lambda = bc$lambda,
    vg = params$vg,
    vm = params$minVM
  )
}

#' Find "best" variogram parameters. 
#' 
#' Find the "best" variogram parameters for a given geodata object. Several
#' different models are tested; the one with the lowest least squares error is
#' chosen.
#' 
#' @section Breaks default:
#' The default breaks argument is set to have more points where the exponential
#' curve rises the most and fewer at large distances. This means that the curve
#' fitting is not overly biased by points beyond the effective maximum range.
#' 
#' @inheritParams geoR::variog
#' 
#' @seealso [geoR::variog].
#' 
#' @return A list of the best fitted variogram parameters. The following
#'   description is adapted from [geoR::variog()] -- which you should see
#'   for more details:
#'   * nugget: value of the nugget parameter. An estimated value if 
#'   `fix.nugget = FALSE` or a fixed value if `fix.nugget = TRUE`.
#'   * sill and range: First and second elements of cov.pars` -- a two elements
#'   vector with estimated values of the covariance parameters sigma^2 and phi,
#'   respectively.
#'   * kappa: Fixed value of the smoothness parameter.
#'   * model: Name of the correlation function (see` cov.model`).
#'   * minVM: The minimum fit error.
#'   * vg: The variogram.
#'
#' @author Graham Zemunik (grah.zem@@gmail.com).
#' 
#' @export
krig_auto_params <- function(geodata,
                             trend = "cte",
                             breaks = krig_breaks(2, 320, 30)) {
  vg <- variog(geodata, breaks = breaks, pairs.min = 5, trend = trend)
  varModels <- c("exponential", "circular", "cauchy", "gaussian")
  minValue <- NULL
  minVM <- NULL
  startRange <- max(breaks) / 2
  initialVal <- max(vg$v) / 2
  for (i in 1:length(varModels)) {
    vm <- variofit(
      vg,
      ini.cov.pars = c(initialVal, startRange),
      nugget = initialVal,
      cov.model = varModels[i]
    )
    if (is.null(minValue) || vm$value < minValue) {
      minValue <- vm$value
      minVM <- vm
    }
  }
  list(
    nugget = minVM$nugget,
    sill = minVM$cov.pars[1],
    range = minVM$cov.pars[2],
    kappa = minVM$kappa,
    model = minVM$cov.model,
    minVM = minVM,
    vg = vg
  )
}

#' Fit a 2D polynomial surface.
#'
#' Fit a 2D polynomial surface from the following inputs:
#' * A dataframe with columns specifying the x, y coordinates and a quantity at
#'   each coord; and
#' * A grid of locations (specified by gridSize), which are interpolated using
#'   nls.
#'
#' @return A list of the original df, the interpolated values at each grid
#'   point, and the nls model, if the nls fit succeeded. No interpolated
#'   locations are returned if nls failed to model the surface, in which case
#'   the model attribute is set to `NULL.`
#'
#' @author Graham Zemunik (grah.zem@@gmail.com).
#'
#' @noRd
#' @keywords internal
GetPolynomialFit <- function(df, gridSize = 20, xSize = 1000, ySize = 500) {
  # The data frame is assumed to be x, y, z
  names(df) <- c("gx", "gy", "z")

  model <- NULL
  tryCatch(
    model <- nls(
      z ~ PolynomialSurfaceOrder2(gx, gy, a, b, c, d, e, f),
      data = df,
      start = list(
        a = 0.1,
        b = 0.1,
        c = 0.1,
        d = 0.1,
        e = 0.1,
        f = 0.1
      ),
      trace = FALSE,
      control = nls.control(maxiter = 200, minFactor = 1 / 4096)
    ),
    error = function(e) {
    }
  )

  halfGrid <- gridSize / 2
  df.locations <- expand.grid(
    gx = seq(halfGrid, xSize - halfGrid, by = gridSize),
    gy = seq(halfGrid, ySize - halfGrid, by = gridSize)
  )
  if (!is.null(model)) {
    df.locations$z <- predict(model, newdata = df.locations)
  }

  list(df.orig = df, df.interpolated = df.locations, mod = model)
}

#' Return a polynomial 2nd order surface (x,y) defined by the parameters a to f.
#'
#' @author Graham Zemunik (grah.zem@@gmail.com).
#'
#' @keywords internal
#' @noRd
PolynomialSurfaceOrder2 <- function(x, y, a, b, c, d, e, f) {
  a + b * x + c * y + d * x * y + e * x^2 + f * y^2
}

#' Find the optimal Box-Cox transform parameters.
#'
#' Finds the optimal Box-Cox transform parameters for the data in data frame,
#' df, with columns specifying the x, y coordinates and a quantity at each
#' coord, whilst restricting the lambda value to 0, 0.5 and 1. Only data >=0 can
#' be transformed. Values = 0 are handled by adding a small value, delta, which
#' if used is returned as the delta argument.
#'
#' @return A list of the original df, the delta value and the the delta value.
#'
#' @author Graham Zemunik (grah.zem@@gmail.com).
#'
#' @keywords internal
#' @noRd
BoxCoxTransformSoil <- function(df) {
  lambda <- 1
  delta <- 0

  if (ncol(df) >= 3) {
    # Sanity checking and enforcement of the structure
    if (!identical(names(df)[1:3], c("gx", "gy", "z"))) {
      names(df)[1:3] <- c("gx", "gy", "z")
    }

    if (min(df$z) >= 0) {
      # boxcox will complain about -ve values
      if (min(df$z) == 0) {
        # add a small amount to allow transforms to work
        delta <- 0.00001
        df$z <- df$z + delta
      }
      bc <- boxcox(
        z ~ gx + gy,
        data = df,
        lambda = c(0, 0.5, 1),
        plotit = F
      )
      lambda <- bc$x[which(bc$y == max(bc$y))]
      lambda <- round(lambda / 0.5) * 0.5 # Get lambda in multiples of 0.5
      if (lambda == 0) {
        df$z <- log(df$z)
      } else {
        if (lambda == 0.5) {
          df$z <- sqrt(df$z)
        }
      }
    }
  }
  # Return the data and the parameters
  list(df = df, lambda = lambda, delta = delta)
}

#' Perform the inverse of the Box-Cox transform.
#'
#' Performed the inverse of the Box-Cox transform given the data, df,
#' the lambda value and and delta added to the data
#'
#' @return The df with the transformed data, from the z column.
#'
#' @author Graham Zemunik (grah.zem@@gmail.com).
#'
#' @keywords internal
#' @noRd
InvBoxCoxTransformSoil <- function(df, lambda, delta) {
  if (lambda == 0) {
    df$z <- exp(df$z)
  } else {
    if (lambda == 0.5) {
      df$z <- df$z^2
    }
  }
  # Take away the delta offset
  df$z <- df$z - delta

  df
}

krig_one <- function(soil,
                 var,
                 params = NULL,
                 gridsize = 20,
                 plotdim = guess_plotdim(soil),
                 breaks = krig_breaks(2, 320, 30),
                 use_ksline = TRUE,
                 quiet = FALSE) {
  krig_msg <- function() {
    message("\nvar: ", var, "Using: gridsize = ", gridsize)
    
    krig_with_message <- enable_quiet(GetKrigedSoil)
    krig <- krig_with_message(
      df.soil = soil,
      var = var,
      gridSize = gridsize,
      krigeParams = params,
      xSize = plotdim[[1]],
      ySize = plotdim[[2]],
      breaks = breaks,
      useKsLine = use_ksline
    )
    
    message(krig$output)
    krig$result
  }
  
  if (quiet) suppressMessages(krig_msg()) else krig_msg() 
}

enable_quiet <- function(krg) {
  force(krg)
  function(...) {
    output <- utils::capture.output({
      result <- krg(...)
    })
    output <- strsplit(paste0(output, collapse = "\n"), "\n\\$df\n")[[1]][1]
    list(
      output = output, 
      result = result
    )
  }
}

check_krig <- function(df.soil,
                       var,
                       gridSize,
                       krigeParams,
                       xSize,
                       ySize,
                       breaks,
                       useKsLine) {
  stopifnot(is.data.frame(df.soil))
  stopifnot(!is.null(df.soil))
  if (!dim(df.soil)[[1]] > 0) {
    stop(
      "df.soil has cero rows\n",
      "  * Ensure `df.soil` has one or more rows",
      call. = FALSE
    )
  }
  fgeo.base::check_crucial_names(df.soil, c("gx", "gy"))
  stopifnot(is.character(var))
  if (length(var) > 1) {
    stop("`var` must be of length 1.\nDo you need `krig()`?")
  }
  stopifnot(!missing(var))

  if (!var %in% names(df.soil)) {
    stop(
      "The variable-name passed to `var` isn't in your data\n",
      "  * Check the possible options with `names(your_data)`",
      call. = FALSE
    )
  }
  stopifnot(is.numeric(gridSize))
  stopifnot(is.numeric(xSize))
  stopifnot(is.numeric(ySize))
  stopifnot(is.numeric(breaks))
  if (!is.null(krigeParams)) {
    stopifnot(is.list(krigeParams))
  }
  stopifnot(is.logical(useKsLine))
}

check_crucial_names <- function(x, nms) {
  are_names_expected <- all(nms %in% names(x))
  if (are_names_expected) {
    return(invisible())
  } else {
    stop(
      "Ensure your data set has these variables:\n",
      paste0(nms, collapse = ", "),
      call. = FALSE
    )
  }
}

new_krig_lst <- function(.x) {
  stopifnot(is.list(.x))
  structure(.x, class = c("krig_lst", class(.x)))
}
